src/conn.ts

```diff
@@ -189,33 +183,45 @@ export class Conn {
   onClientPacket(data: any, meta: PacketMeta, buffer: Buffer, pclient: Client) {
     if (meta.state !== 'play') return;
     const handle = async () => {
-      // Build packet canceler function used by middleware
-      const cancel: PacketCanceler = Object.assign(
-        (unCancel: boolean = false) => {
-          cancel.isCanceled = unCancel ? false : true;
-          update.isUpdated = true;
-        },
-        { isCanceled: false }
-      );
-      const update: packetUpdater = Object.assign(
-        (unUpdate: boolean = false) => {
-          update.isUpdated = !unUpdate;
+      let packetData: PacketData = {
+        bound: 'server', meta, writeType: 'packet', pclient, data: {}, isCanceled: false
+      }
+      let wasChanged = false
+      let isCanceled = false
+      let currentData: PacketData = packetData
+      Object.defineProperties(packetData, {
+        data: {
+          get: () => {
+            wasChanged = true
+            return data
+          }
         },
-        { isUpdated: false }
-      );
+        isCanceled: {
+          get: () => {
+            return isCanceled
+          }
+        }
+      })
 
       for (const middleware of pclient.toServerMiddlewares) {
-        const funcReturn = middleware({ bound: 'server', meta, writeType: 'packet' }, pclient, data, cancel, update);
+        let data: PacketMiddlewareReturnValue
+        const funcReturn = middleware(currentData);
         if (funcReturn instanceof Promise) {
-          await funcReturn;
+          data = await funcReturn;
+        } else {
+          data = funcReturn
+        }
+        isCanceled = data === false
+        if (data !== undefined && data !== false) {
+          currentData = data
         }
       }
-      if (cancel.isCanceled !== false) return;
-      if (!update.isUpdated && this.optimizePacketWrite) {
-        this.writeRaw(buffer);
-        return;
+      if (isCanceled) return
+      if (!wasChanged && this.optimizePacketWrite) {
+        pclient.writeRaw(buffer)
+        return
       }
-      this.write(meta.name, data);
+      pclient.write(meta.name, packetData);
```

Comment:
i don't think this was supposed to be changed

Reply:
at least iirc those two should have quite the different meaning

Reply:
Fixed